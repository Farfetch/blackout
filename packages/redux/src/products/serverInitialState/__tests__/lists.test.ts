import { mockProductsListModel } from 'tests/__fixtures__/products/index.mjs';
import listsServerInitialState from '../listings.js';
import type { Model } from '../../../types/index.js';

describe('products lists serverInitialState()', () => {
  it('should initialize server state for a listing', () => {
    const slug =
      '/en-pt/shopping/woman?pageIndex=1&sort=price&sortDirection=asc';
    // @ts-expect-error A lot of properties would need to be added to make the value comply with the type which are irrelevant for the test
    const model = {
      ...mockProductsListModel,
      slug,
    } as Model;

    const state = listsServerInitialState({ model });

    expect(state).toMatchSnapshot();
  });

  it('should initialize server state for sets', () => {
    const slug = '/en-pt/sets/woman?pageIndex=1&sort=price&sortDirection=asc';
    // @ts-expect-error A lot of properties would need to be added to make the value comply with the type which are irrelevant for the test
    const model = {
      ...mockProductsListModel,
      pageType: 'set',
      slug,
    } as Model;

    const state = listsServerInitialState({ model });

    expect(state).toMatchSnapshot();
  });

  it('should initialize server state for a custom listing page', () => {
    const slug = '/en-pt/customlistingpage';
    // @ts-expect-error A lot of properties would need to be added to make the value comply with the type which are irrelevant for the test
    const model = {
      relatedCommerceData: { referencedListing: [mockProductsListModel] },
      slug,
    } as Model;

    const state = listsServerInitialState({ model });

    expect(state).toMatchSnapshot();
  });

  it('should build the correct sorted hash with encoded query params', () => {
    const slug = '/en-pt/shopping?colors=11%7C6&another=foo';
    const expectedHash = 'listing?colors=11|6&another=foo';
    // @ts-expect-error A lot of properties would need to be added to make the value comply with the type which are irrelevant for the test
    const model = {
      ...mockProductsListModel,
      slug,
    } as Model;

    const state = listsServerInitialState({ model });

    expect(Object.keys(state.entities!.productsLists!)).toEqual(
      expect.arrayContaining([expectedHash]),
    );
    expect(state.lists.hash).toBe(expectedHash);
  });

  it('should build the correct hash without any query params in the slug', () => {
    const slug = '/en-pt/shopping';
    const expectedHash = 'listing';
    // @ts-expect-error A lot of properties would need to be added to make the value comply with the type which are irrelevant for the test
    const model = { ...mockProductsListModel, slug } as Model;
    const state = listsServerInitialState({ model });

    expect(Object.keys(state.entities!.productsLists!)).toEqual(
      expect.arrayContaining([expectedHash]),
    );
    expect(state.lists.hash).toBe(expectedHash);
  });

  it('should build the correct hash when is a custom listing page', () => {
    const slug = '/en-pt/customlistingpage';
    const expectedHash = 'listing/customlistingpage';
    // @ts-expect-error A lot of properties would need to be added to make the value comply with the type which are irrelevant for the test
    const model = {
      relatedCommerceData: { referencedListing: [mockProductsListModel] },
      slug,
    } as Model;
    const state = listsServerInitialState({ model });

    expect(Object.keys(state.entities!.productsLists!)).toEqual(
      expect.arrayContaining([expectedHash]),
    );
    expect(state.lists.hash).toBe(expectedHash);
  });

  it('should initialize server state for a non product list with initial state', () => {
    const model = {} as Model;
    const state = listsServerInitialState({ model });

    expect(state).toMatchSnapshot();
  });

  it('should initialize server state for listing that has returned no products and it is not a custom listing page', () => {
    const slug = '/en-pt/shopping';
    const expectedHash = 'listing';
    const model = {
      dataLayer: { general: { type: 'Listing' } },
      slug,
    } as Model;

    const state = listsServerInitialState({ model });

    expect(state).toMatchSnapshot();
    expect(Object.keys(state.lists.error)).toEqual(
      expect.arrayContaining([expectedHash]),
    );
  });
});
