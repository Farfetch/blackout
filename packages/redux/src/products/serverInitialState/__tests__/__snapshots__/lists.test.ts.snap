// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`products lists serverInitialState() should initialize server state for a listing 1`] = `
Object {
  "entities": Object {
    "facets": Object {
      "categories_144307": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 13,
        "description": "Women",
        "groupsOn": 0,
        "id": "categories_144307",
        "parentId": "categories_0",
        "slug": "women",
        "url": "women",
        "value": 144307,
        "valueUpperBound": 0,
      },
      "categories_144331": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 2,
        "description": "Accessories",
        "groupsOn": 0,
        "id": "categories_144331",
        "parentId": "categories_144307",
        "slug": "women-accessories",
        "url": "women-accessories",
        "value": 144331,
        "valueUpperBound": 0,
      },
      "categories_144424": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 1,
        "description": "Special Accessories",
        "groupsOn": 0,
        "id": "categories_144424",
        "parentId": "categories_144418",
        "slug": "women-specials-special-accessories",
        "url": "women-specials-special-accessories",
        "value": 144424,
        "valueUpperBound": 0,
      },
      "colors_1": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 4,
        "description": "BLACK",
        "groupsOn": 0,
        "id": "colors_1",
        "parentId": "colors_0",
        "slug": null,
        "url": "?colors=1",
        "value": 1,
        "valueUpperBound": 0,
      },
      "pricetype_0": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 13,
        "description": "FullPrice",
        "groupsOn": 0,
        "id": "pricetype_0",
        "parentId": "pricetype_0",
        "slug": "fullprice",
        "url": "fullprice",
        "value": 0,
        "valueUpperBound": 0,
      },
      "sizes_22_144307": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 5,
        "description": "36",
        "groupsOn": 144307,
        "id": "sizes_22_144307",
        "parentId": "sizes_0",
        "slug": null,
        "url": "?sizes=22",
        "value": 22,
        "valueUpperBound": 0,
      },
    },
    "products": Object {
      "123": Object {
        "customAttributes": undefined,
        "groupedEntries": undefined,
        "id": 123,
        "images": undefined,
        "merchant": undefined,
        "price": Object {
          "discount": Object {
            "excludingTaxes": undefined,
            "includingTaxes": undefined,
            "rate": undefined,
          },
          "excludingTaxes": undefined,
          "formatted": Object {
            "includingTaxes": undefined,
            "includingTaxesWithoutDiscount": undefined,
          },
          "includingTaxes": undefined,
          "includingTaxesWithoutDiscount": undefined,
          "isFormatted": true,
          "priceType": undefined,
          "promocode": Object {
            "rate": undefined,
          },
          "promotionType": undefined,
          "tags": undefined,
          "taxes": Object {
            "amount": undefined,
            "rate": undefined,
            "type": undefined,
          },
          "type": undefined,
        },
        "prices": undefined,
        "sizes": undefined,
        "tag": Object {
          "id": undefined,
          "name": undefined,
        },
        "variants": undefined,
      },
      "321": Object {
        "customAttributes": undefined,
        "groupedEntries": undefined,
        "id": 321,
        "images": undefined,
        "merchant": undefined,
        "price": Object {
          "discount": Object {
            "excludingTaxes": undefined,
            "includingTaxes": undefined,
            "rate": undefined,
          },
          "excludingTaxes": undefined,
          "formatted": Object {
            "includingTaxes": undefined,
            "includingTaxesWithoutDiscount": undefined,
          },
          "includingTaxes": undefined,
          "includingTaxesWithoutDiscount": undefined,
          "isFormatted": true,
          "priceType": undefined,
          "promocode": Object {
            "rate": undefined,
          },
          "promotionType": undefined,
          "tags": undefined,
          "taxes": Object {
            "amount": undefined,
            "rate": undefined,
            "type": undefined,
          },
          "type": undefined,
        },
        "prices": undefined,
        "sizes": undefined,
        "tag": Object {
          "id": undefined,
          "name": undefined,
        },
        "variants": undefined,
      },
    },
    "productsLists": Object {
      "listing/woman?pageindex=1&sort=price&sortdirection=asc": Object {
        "breadCrumbs": Array [],
        "config": Object {
          "pageIndex": 1,
        },
        "didYouMean": undefined,
        "facetGroups": Array [
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 1,
            "description": "Categories",
            "dynamic": 0,
            "format": "hierarchical",
            "key": "categories",
            "order": 0,
            "type": 6,
            "values": Array [
              Array [
                "categories_144307",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 2,
            "description": "Categories",
            "dynamic": 0,
            "format": "hierarchical",
            "key": "categories",
            "order": 0,
            "type": 6,
            "values": Array [
              Array [
                "categories_144331",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 3,
            "description": "Categories",
            "dynamic": 0,
            "format": "hierarchical",
            "key": "categories",
            "order": 0,
            "type": 6,
            "values": Array [
              Array [
                "categories_144424",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 0,
            "description": "Colors",
            "dynamic": 0,
            "format": "multiple",
            "key": "colors",
            "order": 5,
            "type": 11,
            "values": Array [
              Array [
                "colors_1",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 0,
            "description": "PriceType",
            "dynamic": 0,
            "format": "multiple",
            "key": "pricetype",
            "order": 6,
            "type": 14,
            "values": Array [
              Array [
                "pricetype_0",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 0,
            "description": "Sizes",
            "dynamic": 0,
            "format": "multiple",
            "key": "sizes",
            "order": 4,
            "type": 9,
            "values": Array [
              Array [
                "sizes_22_144307",
              ],
            ],
          },
        ],
        "facetsBaseUrl": "/en-pt/shopping",
        "filterSegments": Array [
          Object {
            "deep": 1,
            "description": "Women",
            "facetId": "categories_144307",
            "fromQueryString": false,
            "gender": 0,
            "key": "categories",
            "negativeFilter": false,
            "order": 0,
            "parentId": 0,
            "slug": "women",
            "type": 6,
            "value": 144307,
            "valueUpperBound": 0,
          },
        ],
        "gender": 0,
        "genderName": "Woman",
        "hash": "listing/woman?pageindex=1&sort=price&sortdirection=asc",
        "id": 120198,
        "name": "New arrivals",
        "products": Object {
          "entries": Array [
            123,
            321,
          ],
          "totalItems": 193,
          "totalPages": 10,
        },
        "redirectInformation": null,
        "searchTerm": null,
        "slug": "/en-pt/shopping/woman?pageIndex=1&sort=price&sortDirection=asc",
      },
    },
  },
  "lists": Object {
    "error": Object {},
    "hash": "listing/woman?pageindex=1&sort=price&sortdirection=asc",
    "isHydrated": Object {
      "listing/woman?pageindex=1&sort=price&sortdirection=asc": true,
    },
    "isLoading": Object {
      "listing/woman?pageindex=1&sort=price&sortdirection=asc": false,
    },
  },
}
`;

exports[`products lists serverInitialState() should initialize server state for a non products list 1`] = `
Object {
  "lists": Object {
    "error": Object {},
    "hash": null,
    "isHydrated": Object {},
    "isLoading": Object {},
  },
}
`;

exports[`products lists serverInitialState() should initialize server state for sets 1`] = `
Object {
  "entities": Object {
    "facets": Object {
      "categories_144307": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 13,
        "description": "Women",
        "groupsOn": 0,
        "id": "categories_144307",
        "parentId": "categories_0",
        "slug": "women",
        "url": "women",
        "value": 144307,
        "valueUpperBound": 0,
      },
      "categories_144331": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 2,
        "description": "Accessories",
        "groupsOn": 0,
        "id": "categories_144331",
        "parentId": "categories_144307",
        "slug": "women-accessories",
        "url": "women-accessories",
        "value": 144331,
        "valueUpperBound": 0,
      },
      "categories_144424": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 1,
        "description": "Special Accessories",
        "groupsOn": 0,
        "id": "categories_144424",
        "parentId": "categories_144418",
        "slug": "women-specials-special-accessories",
        "url": "women-specials-special-accessories",
        "value": 144424,
        "valueUpperBound": 0,
      },
      "colors_1": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 4,
        "description": "BLACK",
        "groupsOn": 0,
        "id": "colors_1",
        "parentId": "colors_0",
        "slug": null,
        "url": "?colors=1",
        "value": 1,
        "valueUpperBound": 0,
      },
      "pricetype_0": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 13,
        "description": "FullPrice",
        "groupsOn": 0,
        "id": "pricetype_0",
        "parentId": "pricetype_0",
        "slug": "fullprice",
        "url": "fullprice",
        "value": 0,
        "valueUpperBound": 0,
      },
      "sizes_22_144307": Object {
        "_isActive": false,
        "_isDisabled": false,
        "count": 5,
        "description": "36",
        "groupsOn": 144307,
        "id": "sizes_22_144307",
        "parentId": "sizes_0",
        "slug": null,
        "url": "?sizes=22",
        "value": 22,
        "valueUpperBound": 0,
      },
    },
    "products": Object {
      "123": Object {
        "customAttributes": undefined,
        "groupedEntries": undefined,
        "id": 123,
        "images": undefined,
        "merchant": undefined,
        "price": Object {
          "discount": Object {
            "excludingTaxes": undefined,
            "includingTaxes": undefined,
            "rate": undefined,
          },
          "excludingTaxes": undefined,
          "formatted": Object {
            "includingTaxes": undefined,
            "includingTaxesWithoutDiscount": undefined,
          },
          "includingTaxes": undefined,
          "includingTaxesWithoutDiscount": undefined,
          "isFormatted": true,
          "priceType": undefined,
          "promocode": Object {
            "rate": undefined,
          },
          "promotionType": undefined,
          "tags": undefined,
          "taxes": Object {
            "amount": undefined,
            "rate": undefined,
            "type": undefined,
          },
          "type": undefined,
        },
        "prices": undefined,
        "sizes": undefined,
        "tag": Object {
          "id": undefined,
          "name": undefined,
        },
        "variants": undefined,
      },
      "321": Object {
        "customAttributes": undefined,
        "groupedEntries": undefined,
        "id": 321,
        "images": undefined,
        "merchant": undefined,
        "price": Object {
          "discount": Object {
            "excludingTaxes": undefined,
            "includingTaxes": undefined,
            "rate": undefined,
          },
          "excludingTaxes": undefined,
          "formatted": Object {
            "includingTaxes": undefined,
            "includingTaxesWithoutDiscount": undefined,
          },
          "includingTaxes": undefined,
          "includingTaxesWithoutDiscount": undefined,
          "isFormatted": true,
          "priceType": undefined,
          "promocode": Object {
            "rate": undefined,
          },
          "promotionType": undefined,
          "tags": undefined,
          "taxes": Object {
            "amount": undefined,
            "rate": undefined,
            "type": undefined,
          },
          "type": undefined,
        },
        "prices": undefined,
        "sizes": undefined,
        "tag": Object {
          "id": undefined,
          "name": undefined,
        },
        "variants": undefined,
      },
    },
    "productsLists": Object {
      "sets/woman?pageindex=1&sort=price&sortdirection=asc": Object {
        "breadCrumbs": Array [],
        "config": Object {
          "pageIndex": 1,
        },
        "didYouMean": undefined,
        "facetGroups": Array [
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 1,
            "description": "Categories",
            "dynamic": 0,
            "format": "hierarchical",
            "key": "categories",
            "order": 0,
            "type": 6,
            "values": Array [
              Array [
                "categories_144307",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 2,
            "description": "Categories",
            "dynamic": 0,
            "format": "hierarchical",
            "key": "categories",
            "order": 0,
            "type": 6,
            "values": Array [
              Array [
                "categories_144331",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 3,
            "description": "Categories",
            "dynamic": 0,
            "format": "hierarchical",
            "key": "categories",
            "order": 0,
            "type": 6,
            "values": Array [
              Array [
                "categories_144424",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 0,
            "description": "Colors",
            "dynamic": 0,
            "format": "multiple",
            "key": "colors",
            "order": 5,
            "type": 11,
            "values": Array [
              Array [
                "colors_1",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 0,
            "description": "PriceType",
            "dynamic": 0,
            "format": "multiple",
            "key": "pricetype",
            "order": 6,
            "type": 14,
            "values": Array [
              Array [
                "pricetype_0",
              ],
            ],
          },
          Object {
            "_clearUrl": null,
            "_isClearHidden": false,
            "_isClosed": false,
            "deep": 0,
            "description": "Sizes",
            "dynamic": 0,
            "format": "multiple",
            "key": "sizes",
            "order": 4,
            "type": 9,
            "values": Array [
              Array [
                "sizes_22_144307",
              ],
            ],
          },
        ],
        "facetsBaseUrl": "/en-pt/shopping",
        "filterSegments": Array [
          Object {
            "deep": 1,
            "description": "Women",
            "facetId": "categories_144307",
            "fromQueryString": false,
            "gender": 0,
            "key": "categories",
            "negativeFilter": false,
            "order": 0,
            "parentId": 0,
            "slug": "women",
            "type": 6,
            "value": 144307,
            "valueUpperBound": 0,
          },
        ],
        "gender": 0,
        "genderName": "Woman",
        "hash": "sets/woman?pageindex=1&sort=price&sortdirection=asc",
        "id": 120198,
        "name": "New arrivals",
        "products": Object {
          "entries": Array [
            123,
            321,
          ],
          "totalItems": 193,
          "totalPages": 10,
        },
        "redirectInformation": null,
        "searchTerm": null,
        "slug": "/en-pt/sets/woman?pageIndex=1&sort=price&sortDirection=asc",
      },
    },
  },
  "lists": Object {
    "error": Object {},
    "hash": "sets/woman?pageindex=1&sort=price&sortdirection=asc",
    "isHydrated": Object {
      "sets/woman?pageindex=1&sort=price&sortdirection=asc": true,
    },
    "isLoading": Object {
      "sets/woman?pageindex=1&sort=price&sortdirection=asc": false,
    },
  },
}
`;
